libsolidity/formal/CHC.cpp

```diff
@@ -541,6 +541,79 @@ void CHC::endVisit(Return const& _return)
 	m_currentBlock = predicate(*returnGhost);
 }
 
+bool CHC::visit(TryStatement const& _tryStatement)
+{
+	FunctionCall const* externalCall = dynamic_cast<FunctionCall const*>(&_tryStatement.externalCall());
+	solAssert(externalCall && externalCall->annotation().tryCall, "");
+
+	solAssert(m_currentFunction, "");
+	auto const& functionBody = m_currentFunction->body();
+
+	auto tryHeaderBlock = createBlock(&_tryStatement, PredicateType::FunctionBlock, "try_header_");
+	auto tryAfterCallSuccessBlock = createBlock(&_tryStatement, PredicateType::FunctionBlock, "try_after_call_success_");
+	auto tryAfterCallFailBlock = createBlock(&_tryStatement, PredicateType::FunctionBlock, "try_after_call_fail_");
+	auto afterTryBlock = createBlock(&functionBody, PredicateType::FunctionBlock);
+
+	connectBlocks(m_currentBlock, predicate(*tryHeaderBlock));
+	setCurrentBlock(*tryHeaderBlock);
+	// fine-grained control over how the external call is visited
+	// cannot use "_tryStatement.externalCall().accept(*this);"
+	auto const* call = dynamic_cast<FunctionCall const*>(&_tryStatement.externalCall());
+	solAssert(call, "");
+	call->expression().accept(*this);
+	ASTNode::listAccept(call->arguments(), *this);
+	connectBlocks(m_currentBlock, predicate(*tryAfterCallFailBlock));
+	endVisit(*call);
+	if (_tryStatement.successClause()->parameters())
+	{
```

Comment:
This whole block is very similar to what `SMTEncoder::createReturnedExpressions` is doing, but there are small differences.
The whole purpose is to connect the values returned from the external call with the return parameters declared in the `Try` statement.

It would be nice to find some way to do this more elegantly.

Reply:
Yea, this block is pretty much the same as `createReturnedExpressions` without the 3 first lines of that function. I guess you could extract a function with the block below which can also be used by `createReturnedExpressions`.

Reply:
I wanted to do that, but then I notice a small problem, in `createReturnedExpressions` the last line is `defineExpr(_funCall, currentValue(*returnParams.front()));` which assigns the returned values to a *new* version of the expression. However, what I need is the opposite, I need to assign the *current* version of that expression to a variable that I am providing.

I am now thinking more in terms of an assignment, if I can re-use that code somehow.
Because basically `try d.d() returns (uint x, bool b)` is doing the same thing as `(uint x, bool b) = d.d()` (in the success case).

Reply:
@blishko any ideas regarding this block? I still think we should move it. We can also have a call about it.